File: code_samples/Queue/PriorityQueue/PriorityQueue.cpp
// Queue a priorité implémentation en c++

#include <iostream>
#include <vector>
using namespace std;

// Changer place elem pour délétion
void swap(int *a, int *b) {
  int temp = *b;
  *b = *a;
  *a = temp;
}

// "Heapifier" l'abre 
void heapify(vector<int> &hT, int i) {
  int size = hT.size();
  
  // Trouver le plus gros de la racine
  int largest = i;
  int l = 2 * i + 1;
  int r = 2 * i + 2;
  if (l < size && hT[l] > hT[largest])
    largest = l;
  if (r < size && hT[r] > hT[largest])
    largest = r;

  // Effectuer swap et continuer heapify
  if (largest != i) {
    swap(&hT[i], &hT[largest]);
    heapify(hT, largest);
  }
}

// Insertion d'élément
void insert(vector<int> &hT, int newNum) {
  int size = hT.size();
  if (size == 0) {
    hT.push_back(newNum);
  } else {
    hT.push_back(newNum);
    for (int i = size / 2 - 1; i >= 0; i--) {
      heapify(hT, i);
    }
  }
}

// Délétion dans l'abre
void deleteNode(vector<int> &hT, int num) {
  int size = hT.size();
  int i;
  for (i = 0; i < size; i++) {
    if (num == hT[i])
      break;
  }
  swap(&hT[i], &hT[size - 1]);

  hT.pop_back();
  for (int i = size / 2 - 1; i >= 0; i--) {
    heapify(hT, i);
  }
}

// Afficher l'abre
void printArray(vector<int> &hT) {
  for (int i = 0; i < hT.size(); ++i)
    cout << hT[i] << " ";
  cout << "\n";
}

// Main
int main() {
  vector<int> heapTree;

  insert(heapTree, 3);
  insert(heapTree, 4);
  insert(heapTree, 9);
  insert(heapTree, 5);
  insert(heapTree, 2);

  cout << "Array en max-heap: ";
  printArray(heapTree);

  deleteNode(heapTree, 4);

  cout << "Après délétion d'un élément: ";

  printArray(heapTree);
}

File: code_samples/Queue/PriorityQueue/PriorityQueue.py
# Implémentation Queue a priorité en Python


# "Heapifier" l'abre
def heapify(arr, n, i):
    # Trouver le plus gros nombre
    largest = i
    l = 2 * i + 1 #left child
    r = 2 * i + 2 #right child

    if l < n and arr[i] < arr[l]:
        largest = l

    if r < n and arr[largest] < arr[r]:
        largest = r

    # Swap et continuer "Heapifier"
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


# Insertion elem
def insert(array, newNum):
    size = len(array)
    if size == 0:
        array.append(newNum)
    else:
        array.append(newNum)
        for i in range((size // 2) - 1, -1, -1):
            heapify(array, size, i)


# Délétion elem
def deleteNode(array, num):
    size = len(array)
    i = 0
    for i in range(0, size):
        if num == array[i]:
            break

    array[i], array[size - 1] = array[size - 1], array[i]

    array.remove(size - 1)

    for i in range((len(array) // 2) - 1, -1, -1):
        heapify(array, len(array), i)


arr = []

insert(arr, 3)
insert(arr, 4)
insert(arr, 9)
insert(arr, 5)
insert(arr, 2)

print ("Array en max-heap:" + str(arr))

deleteNode(arr, 4)
print("Après délétion d'un élément: " + str(arr))
